firedrake/slate/preconditioners.py

```diff
@@ -0,0 +1,370 @@
+        V_d = FunctionSpace(mesh, broken_elements)
+
+        # Set up the functions for the original, hybridized
+        # and schur complement systems
+        self.broken_solution = Function(V_d)
+        self.broken_rhs = Function(V_d)
+        self.trace_solution = Function(TraceSpace)
+        self.unbroken_solution = Function(V)
+        self.unbroken_rhs = Function(V)
+
+        # Create the symbolic Schur-reduction:
+        # Original mixed operator replaced with "broken"
+        # arguments
+        arg_map = {test: TestFunction(V_d),
+                   trial: TrialFunction(V_d)}
+        Atilde = Tensor(replace(self.cxt.a, arg_map))
+        gammar = TestFunction(TraceSpace)
+        n = ufl.FacetNormal(mesh)
+        sigma = TrialFunctions(V_d)[self.vidx]
+
+        # NOTE: Once extruded is ready, this will change slightly
+        # to include both horizontal and vertical interior facets
+        K = Tensor(gammar('+') * ufl.dot(sigma, n) * ufl.dS)
+
+        # Assemble the Schur complement operator and right-hand side
+        self.schur_rhs = Function(TraceSpace)
+        self._assemble_Srhs = create_assembly_callable(
+            K * Atilde.inv * self.broken_rhs,
+            tensor=self.schur_rhs,
+            form_compiler_parameters=self.cxt.fc_params)
+
+        schur_comp = K * Atilde.inv * K.T
+
+        self.S = allocate_matrix(schur_comp,
+                                 bcs=zero_trace_condition,
+                                 form_compiler_parameters=self.cxt.fc_params)
+        self._assemble_S = create_assembly_callable(
+            schur_comp,
+            tensor=self.S,
+            bcs=zero_trace_condition,
+            form_compiler_parameters=self.cxt.fc_params)
+
+        self._assemble_S()
+        self.S.force_evaluation()
+        Smat = self.S.petscmat
+
+        # Nullspace for the multiplier problem
+        nullspace = create_schur_nullspace(P, K * Atilde.inv,
+                                           V, V_d, TraceSpace,
+                                           pc.comm)
+        if nullspace:
+            Smat.setNullSpace(nullspace)
+
+        # Set up the KSP for the system of Lagrange multipliers
+        ksp = PETSc.KSP().create(comm=pc.comm)
+        ksp.setOptionsPrefix(prefix)
+        ksp.setOperators(Smat)
+        ksp.setUp()
+        ksp.setFromOptions()
+        self.ksp = ksp
+
+        split_mixed_op = dict(split_form(Atilde.form))
+        split_trace_op = dict(split_form(K.form))
+        opts = PETSc.Options()
+        fact_type = opts.getString(prefix + "fieldsplit_schur_fact_type",
+                                   "default")
+
+        # Generate reconstruction calls
+        self._reconstruction_calls(split_mixed_op, split_trace_op, fact_type)
+
+        # Set up the projectors
+        self.data_projector = Projector(self.unbroken_rhs.split()[self.vidx],
+                                        self.broken_rhs.split()[self.vidx])
```

Comment:
I believe it can. Should we allow this to be specified by the user through options? Or just hard-code these into the projector?